localstack/services/s3/codec.py

```diff
@@ -0,0 +1,119 @@
+import io
+from typing import IO, Any, Optional
+
+
+class AwsChunkedDecoder(io.RawIOBase):
+    """
+    This helper class takes a IO[bytes] stream, and decodes it on the fly, so that S3 can directly access the stream
+    without worrying about implementation details of `aws-chunked`.
+    It needs to expose the trailing headers, which will be available once the stream is fully read.
+    You can also directly pass the S3 Object, so the stream would set the checksum value once it's done.
+    TODO: should we register callback instead of passing the object?
+    See `aws-chunked` format here: https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html
+    """
+
+    def readable(self):
+        return True
+
+    def __init__(
+        self, stream: IO[bytes], decoded_content_length: int, s3_object: Optional[Any] = None
+    ):
+        self._stream = stream
+
+        self._decoded_length = decoded_content_length  # Length of the encoded object
```

Comment:
Yeah, we could use some heuristics to determine when the body is over (not having the content length encoded in the chunk), however AWS requires this header and it makes sense to use it to be sure we're properly retrieving the body... but agree it would be nice.